File: 008-StringToInteger/StringToInteger.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
  let res = 0,
    sign = 0,
    index = 0;
  let [negativeLimit, positiveLimit] = [-(2 ** 31), 2 ** 31 - 1];
  while (index < s.length) {
    while (s[index] === " " && index < s.length) index++;
    if (s[index] === "+") {
      if (sign === 0) sign = 1;
      else return 0;
      index++;
    }
    if (s[index] === "-") {
      if (sign === 0) sign = -1;
      else return 0;
      index++;
    }
    if (s[index] >= "0" && s[index] <= "9") {
      while (s[index] >= "0" && s[index] <= "9") {
        res = res * 10 + Number(s[index]);
        index++;
      }
      return sign == -1
        ? Math.max(-res, negativeLimit)
        : Math.min(res, positiveLimit);
    } else return 0;
  }
  return 0;
};

File: 008-StringToInteger/StringToInteger.java

public class StringToInteger {
    public int myAtoi(String s) {
        int current = 0, length = s.length();
        while(current < length && s.charAt(current) == ' '){
            current++;
        }
        if(current >= length) return 0;
        int sign = 0;
        if(s.charAt(current) == '+'){
            sign = 1;
            current++;
        } else if(s.charAt(current) == '-'){
            sign = -1;
            current++;
        }
        int result = 0, negativeLimit = Integer.MIN_VALUE, positiveLimit = Integer.MAX_VALUE;
        while(current < length && s.charAt(current) >= '0' && s.charAt(current) <= '9'){
            int digit = s.charAt(current) - '0';
            if(result > positiveLimit / 10) return sign == -1 ? negativeLimit : positiveLimit;
            if(result == positiveLimit / 10 && digit >= (sign == -1 ? 8: 7)) return sign == -1 ? negativeLimit : positiveLimit;
            result = result * 10 + digit;
            current++;
        }

        return sign == -1 ? -result: result;
    }
    public static void main(String[] args){
        StringToInteger sti = new StringToInteger();
        int value = sti.myAtoi("\" + 431\"");
        System.out.println("value: "+value);
    }
}
